Update env variables docs (#2375)
wasp-lang/wasp: web/docs/project/env-vars.md

While some env vars are required by Wasp, such as the database connection or secrets for social auth, you can also define your env vars for any other useful purposes, and then access them in the code.

Let's go over the available env vars in Wasp, how to define them, and how to use them in your project.

## Client Env Vars

Client environment variables are injected into the client Javascript code during the build process, making them public and readable by anyone. Therefore, you should **never store secrets in them** (such as secret API keys, you should store secrets in the server env variables).

<ClientEnvVarsNote />

You can read them from the client code like this:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title="src/App.js"
import { env } from 'wasp/client'

console.log(env.REACT_APP_SOME_VAR_NAME)
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/App.ts"
import { env } from 'wasp/client'

console.log(env.REACT_APP_SOME_VAR_NAME)
```
</TabItem>
</Tabs>

Read more about the `env` object in the [API reference](#client-env-vars-1).

### Wasp Client Env Vars

Here are the client env vars that Wasp defines:

> #### General Configuration

Reviewer: We haven't introduced earlier this concept of "general configuration env vars" though, right? Should we, or it is self evident?

Author: This is where we introduce it :)

Reviewer: Hm but why did I ask this then? I believe I asked it because you don't really explain it at any moment, you just go for it and start listing them. That is when I was reading it, I was thinking "oh ok general configuration, what is that" and then bam, vars immediately.
Which is why I asked above, should we explain what it is or you think it is self evident - when somebody sees "general configruation", they understand what these env vars are?

Reviewer: I wrote this because when I was reading it I was confused for a bit, so I didn't make it up. Am I representative of a typical user, hard to say. Which is why I asked how you feel about it.

If I would write sometihng, I wouldn't write what you said because it is a bit much, especially the last part with "hence". I might write someting like "Env variables used by Wasp to configure general aspects of your app, not tied to one specific feature.".
But again, is that useful or not -> not sure. I believe I provided my perspective well enough here, I will leave the final decision to you.

Author: Added `These are some general env variables used for various Wasp features:` 👍

<EnvVarsTable>
  <EnvVar name="REACT_APP_API_URL" type="URL" isRequired={false} defaultValue="http://localhost:3001" note="The client uses this as the server URL." />
</EnvVarsTable>

> 

Reviewer: Should we mention that this is it, only one? Somehow it feels like waht now, where is the rest. Maybe we can mention that this is it but you can also deifne your own. Ok maybe that is silly, I am not sure, but this somehow felt unfinished.

Author: It's following the same structure as the server env vars, I think this is understandable.

## Server Env Vars

You can store secret values (e.g. secret API keys) in the server env variables since they are not publicly readable. You can define them without any special prefix, such as `SOME_VAR_NAME=...`.

You can read the env vars from server code like this:
<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js
import { env } from 'wasp/server'

console.log(env.SOME_VAR_NAME)
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts
import { env } from 'wasp/server'

console.log(env.SOME_VAR_NAME)
```
</TabItem>
</Tabs>

Read more about the `env` object in the [API reference](#server-env-vars-1).

### Wasp Server Env Vars

#### General Configuration

<EnvVarsTable>
    <EnvVar name="DATABASE_URL" type="String" isRequired={true} note="The URL of the PostgreSQL database you want your app to use." />

>     <EnvVar name="WASP_WEB_CLIENT_URL" type="URL" isRequired={true} note="Used as your client URL. The server needs to know about it to properly configure Same-Origin Policy (CORS) headers." />

Reviewer: And some other stuff I think, for example provide a link to the client in some automatic emails (auth)? Maybe not any more if flow is completely server-side, but I would maybe state is to that it sound slike CORS is not or might not be the only thing this is important for. I wonder if we should even explain why server needs it -> that info can easily become obsolete, maybe better to skip it? Not a biggie though.

    <EnvVar name="SMTP_PORT" type="Integer" isRequired={true} note="The SMTP server port." />
    <EnvVar name="SMTP_USERNAME" type="String" isRequired={true} note="The SMTP server username." />
    <EnvVar name="SMTP_PASSWORD" type="String" isRequired={true} note="The SMTP server password." />
</EnvVarsTable>

#### SendGrid Email Sender

If you are using `SendGrid` as your email sender, you need to provide the following environment variables:

<EnvVarsTable>
    <EnvVar name="SENDGRID_API_KEY" type="String" isRequired={true} note="The SendGrid API key." />
</EnvVarsTable>

#### Mailgun Email Sender

If you are using `Mailgun` as your email sender, you need to provide the following environment variables:

<EnvVarsTable>
    <EnvVar name="MAILGUN_API_KEY" type="String" isRequired={true} note="The Mailgun API key." />
    <EnvVar name="MAILGUN_DOMAIN" type="String" isRequired={true} note="The Mailgun domain." />
    <EnvVar name="MAILGUN_API_URL" type="URL" isRequired={false} note={
      <span>Useful if you want to use the EU API endpoint (<code>https://api.eu.mailgun.net</code>).</span>
    } />
</EnvVarsTable>

#### Auth

> You can use this env while developing to skip email verification. This is useful when you are testing the email sending functionality in development:

Reviewer: You are talking here about a single specific env var? That is a bit weird because below we have a whole table. Feels like you are talking about the whole table. Maybe put it all in the note for that one instead?

Reviewer: Did this ever happen (I don't see a rocket :rocket: )?

Author: Yep, it happened, sorry 🙋‍♂️

> You can use this env while developing to skip email verification. This is useful when you are testing the email sending functionality in development:

Reviewer: ```suggestion
You can use this env var during development to skip email verification. This is useful when you are testing the email sending functionality in development:
```
The part "this is useful when you are testing the email sending functionality in development" also confuses me -> how is this useful if I am testing the email sending functionality? I guess it is useful if I am not testing email sending functinoality? Hmmmm

> <!-- TODO: this not really Boolean, it's really a string that accepts "true" as only true value -->

Reviewer: Ay yay let's make that clear?

Author: I'll update this once we merge the env vars PR since the type depends on what we decide we want to do there. #2362

Reviewer: Ok, as long as you don't forget about it!

<EnvVarsTable>
    <EnvVar name="SKIP_EMAIL_VERIFICATION_IN_DEV" type="Boolean" isRequired={false} defaultValue="false" note="If set to true, automatically sets user emails as verified in development." />
</EnvVarsTable>

#### Jobs

<EnvVarsTable>
    <EnvVar name="PG_BOSS_NEW_OPTIONS" type="String" isRequired={false} note={
        <span>It's parsed as JSON. Enables you to provide <a href="../advanced/jobs#declaring-jobs">custom config</a> for PgBoss.</span>
    } />
</EnvVarsTable>

> #### OAuth Configuration

Reviewer: Can we put this under Auth header?

Author: I've moved the `Auth` stuff to `Development` and now this feels better i.e. I named it `OAuth providers`.

If you are using OAuth, you need to provide the following environment variables:

<EnvVarsTable>
    <EnvVar name="<PROVIDER_NAME>_CLIENT_ID" type="String" isRequired={true} note="The client ID provided by the OAuth provider." />
    <EnvVar name="<PROVIDER_NAME>_CLIENT_SECRET" type="String" isRequired={true} note="The client secret provided by the OAuth provider." />
</EnvVarsTable>

 `<PROVIDER_NAME>` is the uppercase name of the provider you are using. For example, if you are using Google OAuth, you need to provide the `GOOGLE_CLIENT_ID` and `GOOGLE_CLIENT_SECRET` environment variables.

> If you are using [Keycloak](../auth/social-auth/keycloak.md), you'll need to provide one extra environment variable:

Reviewer: Should we create special header for Keycloak?

Author: I feel like it would confuse people, just keeping it under `OAuth Providers` makes it feel connected to the rest of the stuff.

    REACT_APP_SOME_VAR_NAME=somevalue
    ```

    <ClientEnvVarsNote />

`.env.server` should not be committed to version control as it can contain secrets, while `.env.client` can be versioned as it must not contain any secrets.
By default, in the `.gitignore` file that comes with a new Wasp app, we ignore all dotenv files.

> :::info Dotenv files

Reviewer: Where did this go, it wasn't useful?

Author: It didn't feel useful, it talks about the specific Node.js package and they are not going to use it directly. I think the env vars file format is self-explanatory.

Reviewer: They might need to understand how to use it though, to for example pass JSON values and similar, we had those situations, how to quote stuff. So I thoguht it is good to introduce the concept, since we are using it as na important part of Wasp, so it is clear this is not something we made up, but instead there are more resources behind it that they can look for. Drop it if you really want, but to me it seems useful to keep it.

Author: I get what you mean, if people know what's underneath they can Google better. 

The env variables are first parsed by Haskell and then outputted to the `.env` files in the server and the client apps. If we wanted to provide full information, we should first mention Haskell's [dotenv](https://hackage.haskell.org/package/dotenv) package and then our logic for outputting the vars i.e. our `envVarsToDotEnvContent` function. And then we can mention that the variables are ingested by Node.js with the JS [dotenv](https://www.npmjs.com/package/dotenv) package.

I dislike giving this many details to users because 99% of them don't really need them and those with issues will most likely ask about it in our Discord.

## Defining Env Vars in Production

> While in development, we had the option of using `.env.client` and `.env.server` files which made it easy to define and manage env vars.

Reviewer: Aha this went away because it is now mentinoed in deployment/env-vars.md?

Author: I guess so, yes 😄 I moved things around so it made sense when reading the docs.

> We'll talk about how to define env vars for each deployment option in the [deployment section](../deployment/env-vars.md).

Reviewer: We'll talk -> why is it future tense?

We'll talk about how to define env vars for each deployment option in the [deployment section](../deployment/env-vars.md).

## Custom Env Var Validations

TODO: when the Zod validation PRs are merged, describe how users can define their own validations.

> ![Env vars usage in development and production](/img/env/prod_dev_fade_2.svg)

Reviewer: What happened to this image?

Author: It moved to the deployment env vars page :)
